curvefs/src/client/fuse_client.cpp

```diff
@@ -946,68 +948,154 @@ CURVEFS_ERROR FuseClient::CalOneLayerSumInfo(Inode *inode) {
+        } else {
+            {
+                std::lock_guard<std::mutex> guard(*stackMutex);
+                for (const auto &it : dentryList) {
                     if (it.type() == FsFileType::TYPE_DIRECTORY) {
-                        rsubdirs++;
-                    } else {
-                        rfiles++;
+                        iStack->emplace(it.inodeid());
                     }
-                    rentries++;
-                    rfbytes += it.length();
-                    // record hardlink
-                    if (it.type() != FsFileType::TYPE_DIRECTORY &&
-                        it.nlink() > 1) {
-                        if (hardLinkMap.count(it.inodeid())) {
-                            hardLinkMap[it.inodeid()] += it.length();
+                    inodeIds.emplace(it.inodeid());
+                }
+            }
+            inflightNum->fetch_sub(1, std::memory_order_seq_cst);
+            if (!inodeIds.empty()) {
+                auto tret = inodeManager_->BatchGetInodeAttr(&inodeIds, &attrs);
+                if (tret == CURVEFS_ERROR::OK) {
+                    std::lock_guard<std::mutex> guard(*valueMutex);
+                    for (const auto &it : attrs) {
+                        if (it.type() == FsFileType::TYPE_DIRECTORY) {
+                            (*rsubdirs)++;
                         } else {
-                            hardLinkMap.emplace(it.inodeid(), 0);
+                            (*rfiles)++;
+                        }
+                        (*rentries)++;
+                        (*rfbytes) += it.length();
+                        // record hardlink
+                        if (it.type() != FsFileType::TYPE_DIRECTORY &&
+                            it.nlink() > 1) {
+                            std::lock_guard<std::mutex> guard(*mapMutex);
+                            auto iter = hardLinkMap->find(it.inodeid());
+                            if (iter != hardLinkMap->end()) {
+                                iter->second += it.length();
+                            } else {
+                                hardLinkMap->emplace(it.inodeid(), 0);
+                            }
                         }
                     }
+                } else {
+                    ret->exchange(false);
+                    break;
                 }
-                inodeIds.clear();
-                attrs.clear();
-            } else {
-                return ret;
             }
         }
     }
+}
+
+CURVEFS_ERROR FuseClient::CalAllLayerSumInfo(Inode *inode) {
+    std::stack<uint64_t> iStack;
+    std::mutex stackMutex;
+
+    // record hard link, <inodeId, need2minus>
+    std::unordered_map<uint64_t, uint64_t> hardLinkMap;
+    std::mutex mapMutex;
+
+    uint64_t rfiles = 0;
+    uint64_t rsubdirs = 0;
+    uint64_t rentries = 0;
+    uint64_t rfbytes = 0;
+    std::mutex valueMutex;
+
+    auto ino = inode->inodeid();
+    iStack.emplace(ino);
+    std::vector<Thread> threadpool;
```

Comment:
you should take care of the dynamic threadpool, if there're many requests, it will start many many threads, but the total number of threads that a process can start is limited, and if the limit is reached, creating a new thread will throw an exception which will lead to abort the process.

Reply:
yeap, using try and catch block to deal the failure when creat thread. It is doesn't matter if the thread num created less the wanted.